vulkan/vulkan_image.cc

```diff
@@ -0,0 +1,72 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/vulkan/vulkan_command_buffer.h"
+#include "flutter/vulkan/vulkan_image.h"
+#include "flutter/vulkan/vulkan_proc_table.h"
+
+namespace vulkan {
+
+VulkanImage::VulkanImage(VulkanHandle<VkImage> image)
+    : handle_(std::move(image)),
+      layout_(VK_IMAGE_LAYOUT_UNDEFINED),
+      access_flags_(0),
+      valid_(false) {
+  if (!handle_) {
+    return;
+  }
+
+  valid_ = true;
+}
+
+VulkanImage::~VulkanImage() = default;
+
+bool VulkanImage::IsValid() const {
+  return valid_;
+}
+
+const VulkanHandle<VkImage>& VulkanImage::Handle() const {
+  return handle_;
+}
+
+bool VulkanImage::InsertImageMemoryBarrier(
+    const VulkanCommandBuffer& command_buffer,
+    VkPipelineStageFlagBits src_pipline_bits,
+    VkPipelineStageFlagBits dest_pipline_bits,
+    VkAccessFlagBits dest_access_flags,
+    VkImageLayout dest_layout) {
+  const VkImageMemoryBarrier image_memory_barrier = {
+      .sType = VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER,
+      .pNext = nullptr,
+      .srcAccessMask = access_flags_,
+      .dstAccessMask = dest_access_flags,
+      .oldLayout = layout_,
+      .newLayout = dest_layout,
+      .srcQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED,
+      .dstQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED,
+      .image = handle_,
+      .subresourceRange = {VK_IMAGE_ASPECT_COLOR_BIT, 0, 1, 0, 1},
+  };
+
+  bool success = command_buffer.InsertPipelineBarrier(
+      src_pipline_bits,      // src_stage_flags
+      dest_pipline_bits,     // dest_stage_flags
+      0,                     // dependency_flags
```

Comment:
Are you sure that you don't want VK_DEPENDENCY_BY_REGION_BIT?  I'm not 100% clear on the semantics (I have a mental TODO about it), but I've set image memory-barriers that didn't work properly without this bit set.